proposals/collection-literals.md

```diff
@@ -151,70 +156,76 @@ _Open issue: Relying on a parameter named `capacity` seems brittle. Is there an
+The *create method* must be a `static` method defined on the *builder type*.
+The *create method* must have a single parameter of a type `System.ReadOnlySpan<T0>` passed by value, and a return value of the *collection type* to which the attribute is applied.
 
-    ```c#
-    T[] __storage = /* initialized using predefined rules */
-    ImmutableArray<T> __result = new ImmutableArray<T>();
-    __result.Construct(__storage);
-    ```
+The *builder type* cannot be generic.
+The *builder method* can be a generic *method* and the *arity* of the builder method must match the *arity* of the *collection type*.
 
-    `ImmutableArray<T>` would then take that array directly and use it as its own backing storage.  This would be safe because the compiler (following the requirements around `init`) would ensure that no other location in the code would have access to this temporary array, and thus it would not be possible to mutate it behind the back of the `ImmutableArray<T>` instance.
+Overloads with distinct signatures on the builder type, and overloads on base types, are ignored.
 
-    The above also demonstrates that this approach can work with struct types which do not have a [*parameterless struct constructor*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/parameterless-struct-constructors.md).  In the above, the call to `new ImmutableArray<T>()` is equivalent to `default(ImmutableArray<T>)`, (producing an `ImmutableArray<T>` whose [`IsDefault`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablearray-1.isdefault) property is initially true.  However, the `Construct` method can then safely update this to the final non-default state without that intermediate state being visible.
+An error is reported for a collection literal if the `[CollectionBuilder]` attribute on the collection type does not represent an invocable method with the expected signature defined on the *builder type*.
 
-* This formalization is quite beneficial because the only existing mechanism to (safely) create an ImmutableArray with values without copying is both excessively verbose and produces unavoidable garbage:
+If the parameter does not escape the *create method* (that is, if the parameter is `scoped` or if the return type is not a `ref struct`), the compiler *may* allocate the storage for the `ReadOnlySpan<T0>` on the stack rather than the heap.
 
-    ```c#
-    var __builder = ImmutableArray.CreateBuilder<int>(initialCapacity: __len);
+For example, a possible *create method* for `ImmutableArray<T>`:
+```csharp
+[CollectionBuilder(typeof(ImmutableArray), "Create")]
+public struct ImmutableArray<T> { ... }
 
-    __builder.Add(e1);
-    foreach (var __t in s1)
-        __builder.Add(__t);
+public static class ImmutableArray
+{
+    public static ImmutableArray<T> Create<T>(ReadOnlySpan<T> items) { ... }
```

Comment:
```suggestion
    public static ImmutableArray<T> Create<T>(scoped ReadOnlySpan<T> items) { ... }
```

Ideally this parameter is `scoped`

Reply:
I had this as `scoped` initially but removed the modifier from this example since it's redundant.